File: Apps/Pecunia/App/Finance/ContactsPanel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using Codaxy.Dextop;
using Codaxy.Dextop.Remoting;
using Codaxy.Dextop.Data;
using System.Xml.Linq;
using System.Globalization;
using System.Xml;
using Codaxy.Dextop.Forms;
using Pecunia.Services;
using Codaxy.Common;

namespace Pecunia.App
{
    public class ContactsPanel : DextopWindow
    {
        [DextopRemotableConstructor(alias = "contacts")]
        public ContactsPanel()
        {
        }

		public override void InitRemotable(DextopRemote remote, DextopConfig config)
		{
			base.InitRemotable(remote, config);
			Remote.AddStore("model", new ContactCrud());
		}
    }

	class ContactCrud : IDextopDataProxy<Contact>
	{
		Dictionary<String, Contact> Data { get; set; }

		public ContactCrud()
		{
			Data = new Dictionary<string, Contact>();
			foreach (var c in ContactService.GetContacts())
				Data[c.Id] = c;
		}

		public IList<Contact> Create(IList<Contact> records)
		{
			foreach (var rec in records)
			{
				rec.Id = Guid.NewGuid().ToString();
				Data.Add(rec.Id, rec);
			}
			ContactService.SaveContacts(Data.Values.ToArray());
			return records;
		}

		public IList<Contact> Destroy(IList<Contact> records)
		{
			foreach (var rec in records)
			{
				Data.Remove(rec.Id);
			}
			ContactService.SaveContacts(Data.Values.ToArray());
			return new Contact[0];
		}

		public IList<Contact> Update(IList<Contact> records)
		{
			foreach (var rec in records)
			{
				Data[rec.Id] = rec;
			}
			ContactService.SaveContacts(Data.Values.ToArray());
			return records;
		}

		public DextopReadResult<Contact> Read(DextopReadFilter filter)
		{
			return DextopReadResult.Create(Data.Values);
		}
	}

    [DextopForm]
    [DextopModel]
    [DextopGrid]
    public class Contact
    {
        [DextopModelId]
        public string Id { get; set; }

		[DextopFormField(anchor = "0")]
		[DextopGridColumn(flex = 1, text = "Name")]
		public String Name { get; set; }

		[DextopFormTabPanel(itemId="tab")]
		[DextopFormPanel(1, title = "General", bodyStyle = "padding: 5px", layout = "anchor")]
        [DextopFormFieldSet(2, title = "Personal")]
        [DextopFormField()]        
        public String DOB { get; set; }

		[DextopFormField(fieldLabel="Photo URL", anchor="0")]        
		public String PhotoUrl { get; set; }

		[DextopFormField(fieldLabel = "Wikipedia URL", anchor = "0")]
		public String WikipediaUrl { get; set; }

		[DextopFormField(anchor="0")]
		public String Nationality { get; set; }

		[DextopFormField(anchor = "0")]
		public string Religion { get; set; }

		[DextopFormField()]
		public int? Children { get; set; }

		[DextopFormFieldSet(2, title = "Life")]        
		[DextopFormField(anchor = "0")]        
        public String Business { get; set; }

        [DextopFormField(fieldLabel = "Fortune US$ (billion)")]
        [DextopGridColumn(width = 100)]
        public double Fortune { get; set; }

		[DextopFormField(anchor = "0")]
		public string Occupation { get; set; }

		[DextopFormPanel(1, title = "Bio", bodyStyle = "padding: 5px", layout = "anchor")]
		[DextopFormTextArea(anchor="0", height=200, labelAlign="top")]
		public string Bio { get; set; }

		[DextopFormPanel(1, title = "Extra", bodyStyle = "padding: 5px", layout="anchor")]
		[DextopFormField(anchor = "0", fieldLabel="Title", labelAlign="top")]
		public string ExtraTitle { get; set; }

		[DextopFormTextArea(anchor = "0", height = 200, labelAlign = "top")]
		public string Extra { get; set; }


	}

}

File: Apps/Pecunia/App/Finance/ContactsPanel.js
Ext.define('Pecunia.ContactsPanel', {
	extend: 'Dextop.Panel',

	title: 'Contacts',
	border: true,
	closable: true,

	initComponent: function () {

		var preview = Ext.widget('panel', {
			autoScroll: true,
			region: 'center',
			bodyStyle: 'padding-top: 5px',
			border: false,
			tpl: new Ext.XTemplate('<div class="person">',
			'<div style="float: left; padding: 0 5px 0px 0;">{[this.photo(values)]}</div>',
			'<h1>{Name}</h1>',
			'<p><i>Source:</i> <a href="{WikipediaUrl}" target="_blank">Wikipedia</a></p>',
			'<table>',
			'<tr><td class="prop-header">Fortune:</td><td class="prop-value">US$ {Fortune} billion</td></tr>',
			'<tr><td class="prop-header">Occupation:</td><td class="prop-value">{[this.occupation(values.Occupation)]}</td></tr>',
			'<tr><td class="prop-header">DOB:</td><td class="prop-value">{DOB}</td></tr>',
			'<tr><td class="prop-header">Nationality:</td><td class="prop-value">{Nationality}</td></tr>',
			'<tr><td class="prop-header">Religion:</td><td class="prop-value">{Religion}</td></tr>',
			'<tr><td class="prop-header">Children:</td><td class="prop-value">{Children}</td></tr>',
			'</table>',
			'<h2>About</h2>',
			'<p>{Bio}</p>',
			'<h2>{ExtraTitle}</h2>',
			'<p>{Extra}</p>',
			'</div>', {
				photo: function (values) {
					if (values.PhotoUrl)
						return '<image class="photo" src="' + values.PhotoUrl + '" />';
					return '';
				},
				occupation: function (value) {
					if (!value)
						return value;
					return value.replace(new RegExp(';', 'g'), '<br/>');
				}
			})
		});

		var grid = Ext.create('Dextop.ux.SwissArmyGrid', {
			region: 'west',
			width: 400,
			remote: this.remote,
			border: true,
			tbar: ['add', 'edit', 'remove'],
			margins: '5',
			paging: false,
			editing: 'form',
			storeOptions: {
				autoLoad: true,
				autoSync: true,
				sortOnLoad: true,
				sorters: [{ property: 'Fortune', direction: 'DESC' }]
			},
			listeners: {
				scope: this,
				select: function (sm, record) {
					preview.update(record.data);
				}
			},
			editingOptions: {
				width: 600,				
				formItemsConfig: {
					apply: {
						'tab': {
							plain: true
						}
					}
				}
			}
		});

		Ext.apply(this, {
			layout: 'border',
			items: [grid, preview]
		});

		this.callParent(arguments);

	}
});